### DISCUSSION TITLE
What should be taken into considiration if I want to impliment "PutBucketLifecycleConfigurationHandler" API

### DISCUSSION THREAD

**Discussion by guol-fnst** [seaweedfs/seaweedfs#2703] [Ideas]
At first，I want to impliment simple lifecycle policy such as setting  TTL for bucket.
What should be taken into considiration if I want to impliment "PutBucketLifecycleConfigurationHandler" and "PutBucketLifecycle" API.

**Comment by guol-fnst** [comment]
According to https://docs.aws.amazon.com/AmazonS3/latest/API/API_LifecycleRule.html#AmazonS3-Type-LifecycleRule-Expiration
At the first step:
**supported rules:**
- Expiration
- Filter
- ID
- Status

**unsupported rules:**
- AbortIncompleteMultipartUpload
- NoncurrentVersionExpiration
- Transitions

**Comment by chrislusf** [comment]
we need a mechanism to execute these rules first.

**Reply by guol-fnst** [reply]
1. add new data struct  FilerConf_PathConf.BucketRules   
```
type FilerConf_PathConf struct {
	state         protoimpl.MessageState
	sizeCache     protoimpl.SizeCache
	unknownFields protoimpl.UnknownFields

	LocationPrefix    string                           `protobuf:"bytes,1,opt,name=location_prefix,json=locationPrefix,proto3" json:"location_prefix,omitempty"`
	Collection        string                           `protobuf:"bytes,2,opt,name=collection,proto3" json:"collection,omitempty"`
	Replication       string                           `protobuf:"bytes,3,opt,name=replication,proto3" json:"replication,omitempty"`
	Ttl               string                           `protobuf:"bytes,4,opt,name=ttl,proto3" json:"ttl,omitempty"`
	DiskType          string                           `protobuf:"bytes,5,opt,name=disk_type,json=diskType,proto3" json:"disk_type,omitempty"`
	Fsync             bool                             `protobuf:"varint,6,opt,name=fsync,proto3" json:"fsync,omitempty"`
	VolumeGrowthCount uint32                           `protobuf:"varint,7,opt,name=volume_growth_count,json=volumeGrowthCount,proto3" json:"volume_growth_count,omitempty"`
	ReadOnly          bool                             `protobuf:"varint,8,opt,name=read_only,json=readOnly,proto3" json:"read_only,omitempty"`
	DataCenter        string                           `protobuf:"bytes,9,opt,name=data_center,json=dataCenter,proto3" json:"data_center,omitempty"`
	Rack              string                           `protobuf:"bytes,10,opt,name=rack,proto3" json:"rack,omitempty"`
	DataNode          string                           `protobuf:"bytes,11,opt,name=data_node,json=dataNode,proto3" json:"data_node,omitempty"`
	BucketRules       []*FilerConf_PathConf_BucketRule `protobuf:"bytes,12,rep,name=bucket_rules,json=bucketRules,proto3" json:"bucket_rules,omitempty"`
}
```
    
2. Implement s3 api "PutBucketLifecycleConfigurationHandler" to write  lifecycle rules to filer confiigure files.
3. Select a filer to run   goroutine " process_bucket_lifecycle" , which checks bucket lifecycle rules and delete matched buckets or bojects. 
 I tried to find a perfect method to choose a proper filer to process lifecycle rules. The mechanism must be highly available and mutually exclusive. If the lifecycle-filer crashes or exits, another filer must start processing lifecycle rules continuely.  And at the same time ,there should be only one filer to process lifecycle rules.
At first, I think using raft to choose a master filer as the  lifecycle-filer. But it seems to be a little bit complicated.
Then I find a new method. The following functions monitor the events of all filer clients. All alive filer clients can be  saved in a list, and master sends grpc to the first filer client in the list to process lifecyle rules. If this lifecycle-filer client crashes, master will be notified and this lifecycle-filer client will be removed from the list, then a new first filer client in the list will process lifecycle continuely.
```
func (ms *MasterServer) addClient
func (ms *MasterServer) deleteClient
```
If there are any problems about this mechanism please let me know. I am very pleased to contribute PR to seaweedfs.

**Reply by chrislusf** [reply]
"checks bucket lifecycle rules and delete matched buckets or objects."

Checking files one by one would not scale.

**Reply by guol-fnst** [reply]
"Checking files one by one would not scale."
I thinks "Checking files one by one" only happens when objecs tags are specified in a filter rule.  A example XML is as follows.
```
<LifecycleConfiguration>
  <Rule>
    <ID>Rule 1</ID>
    <Filter>
      <Tag>
         <Key>tag1</Key>
         <Value>value1</Value>
      </Tag>
    </Filter>
    <Status>Enabled</Status>
    <Transition>
      <StorageClass>GLACIER<StorageClass>
      <Days>365</Days> 
    </Transition>
  </Rule>
</LifecycleConfiguration>
```
In the current filer architecture, I couldn't figure out a elegant way to process rules for tagged object.
Maybe adding a bool key to datebase associate with entry to indicate a tagged object, which will speed up processing rules when scanning objects. But it 's still a inefficient method. Could you provide some helpful ideas?



In other situations, we just need to delete a bucket or a sub directory(when a prefix is specified in a rule) in a bucket.

https://docs.aws.amazon.com/AmazonS3/latest/userguide/intro-lifecycle-rules.html#intro-lifecycle-rules-filter
```
Filter element
A Lifecycle rule can apply to all or a subset of objects in a bucket based on the <Filter> element that you specify in the Lifecycle rule.

You can filter objects by key prefix, object tags, or a combination of both (in which case Amazon S3 uses a logical AND to combine the filters). Consider the following examples:
```

**Reply by guol-fnst** [reply]
> "checks bucket lifecycle rules and delete matched buckets or objects."
> 
> Checking files one by one would not scale.

Besides do you have any advice on choosing a filer to process lifecycle?

**Reply by guol-fnst** [reply]
Update about choosing lifecycle-filer :
step 1: On startup, Master leader checks all alive filer, if no filer is processing lifecycle rules, choose the first one of the filer list.
step 2: If the filer which is processing lifecycle rules crashes, it should be delete from filer list. Then master leader  chooses the new first on as the  lifecycle-filer
step 3: If the master leader crashes, another node will be selected  to be leader. Then go to step 1.

**Reply by gada121982** [reply]
Any update on this feature @guol-fnst? I already developed this feature and i want to contribute to seaweedFS. We can work on this feature together, what do you think @guol-fnst.
